lib/galaxy/jobs/runners/tes.py

```diff
@@ -0,0 +1,547 @@
+
+        script[-1] = '\n'
+        return ' '.join(script)
+
+    def staging_out_command(self, script_path: str, output_files: list, api_url: str, work_dir: str):
+        """
+            Generates command for staging out of files
+        """
+        command = ["python", script_path, api_url, work_dir]
+
+        for file in output_files:
+            command.append(file)
+
+        return command
+
+    def env_variables(self, job_wrapper: JobWrapper):
+        """
+            Get environment variables from job_wrapper
+        """
+        env_vars = {}
+        for variable in job_wrapper.environment_variables:
+            env_vars[variable['name']] = variable['value']
+
+        for variable in job_wrapper.job_destination.env:
+            env_vars[variable['name']] = variable['value']
+
+        env_vars["_GALAXY_JOB_TMP_DIR"] = self.container_workdir
+        env_vars["_GALAXY_JOB_HOME_DIR"] = self.container_workdir
+        return env_vars
+
+    def input_url(self, api_url: str, path: str):
+        """
+            Get URL for path
+        """
+        file_link = f"{api_url}&path={path}"
+        return file_link
+
+    def input_descriptors(self, api_url: str, input_paths: list, type: str = "FILE"):
+        """
+            Get Input Descriptor for Jobfile
+        """
+        input_description = []
+
+        for path in input_paths:
+            input_description.append({
+                "url": self.input_url(api_url, path),
+                "path": path,
+                "type": type
+            })
+        return input_description
+
+    def get_job_directory_files(self, work_dir: str):
+        """
+        Get path for all the files from work directory
+        """
+        paths = []
+        for root, _, files in os.walk(work_dir):
+            for file in files:
+                paths.append(os.path.join(root, file))
+
+        return paths
+
+    def get_docker_image(self, job_wrapper: JobWrapper):
+        """
+            For getting the docker image to be used
+        """
+        remote_container = self._find_container(job_wrapper)
+        remote_image = "vipchhabra99/pulsar-lib"
+
+        if(hasattr(job_wrapper.job_destination, "params")):
+            if("default_docker_image" in job_wrapper.job_destination.params):
+                remote_image = job_wrapper.job_destination.params.get("default_docker_image")
+
+        if(hasattr(remote_container, "container_id")):
+            remote_image = remote_container.container_id
+
+        return remote_image
+
+    def build_script(self, job_wrapper: JobWrapper, client_args: dict):
+        """
+        Returns the Job script with required configurations of the job
+        """
+        tool_dir = job_wrapper.tool.tool_dir
+        work_dir = job_wrapper.working_directory
+        object_store_path = job_wrapper.object_store.file_path
+        mount_path = os.path.commonprefix([work_dir, object_store_path])
```

Comment:
I will have to look into this. I am not really sure about it.

Reply:
> > Hi, I have written a report explaining the overall process https://gist.github.com/vipulchhabra99/ad41343524cb6ba5d27f57098fca0c58. It just needs setting up a TES implementation and then setting up the config files as shown in the video https://www.youtube.com/watch?v=cAbgGSXu9_g. It doesn't require a shared file system, and it pushes the file back using Galaxy API. There could be some delays in replying and fixing all the suggestions really sorry for that.
> 
> Thanks for working on this. Can you please write some documentation on how to set this up? We cannot maintain or update videos as documentation. A good place for this would be a new section for TES following https://github.com/galaxyproject/galaxy/blob/dev/doc/source/admin/cluster.md#pulsar.
> 
> Please also mention the shortcomings of this approach as you see them.

Sure, I will update it, in few days

Reply:
hopefully, it's fixed now